vhost-device-spi/src/spi.rs

```diff
@@ -0,0 +1,1119 @@
+const CONFIG_SPACE_CPOL_0: u32 = 0x4;
+const CONFIG_SPACE_CPOL_1: u32 = 0x8;
+const CONFIG_SPACE_CS_HIGH: u32 = 0x10;
+const CONFIG_SPACE_LSB: u32 = 0x20;
+const CONFIG_SPACE_LOOP: u32 = 0x40;
+
+/// Mode setting in Requests
+pub const SPI_CPHA: u32 = 1 << 0;
+pub const SPI_CPOL: u32 = 1 << 1;
+pub const SPI_CS_HIGH: u32 = 1 << 2;
+pub const SPI_LSB_FIRST: u32 = 1 << 3;
+pub const SPI_LOOP: u32 = 1 << 4;
+
+/// Copied (partially) from Linux's include/uapi/linux/spi/spidev.h
+///
+/// SpiIocTransfer - describes a single SPI transfer
+///
+/// @tx_buf: Holds pointer to userspace buffer with transmit data, or null.
+///
+/// @rx_buf: Holds pointer to userspace buffer for receive data, or null.
+///
+/// @len: Length of tx and rx buffers, in bytes.
+///
+/// @speed_hz: Temporary override of the device's bitrate.
+///
+/// @bits_per_word: Temporary override of the device's wordsize.
+///
+/// @delay_usecs: If nonzero, how long to delay after the last bit transfer
+///   before optionally deselecting the device before the next transfer.
+///
+/// @cs_change: True to deselect device before starting the next transfer.
+///
+/// @word_delay_usecs: If nonzero, how long to wait between words within one
+///   transfer. This property needs explicit support in the SPI controller,
+///   otherwise it is silently ignored.
+///
+/// This structure is mapped directly to the kernel spi_transfer structure;
+/// the fields have the same meanings, except of course that the pointers
+/// are in a different address space (and may be of different sizes in some
+/// cases, such as 32-bit i386 userspace over a 64-bit x86_64 kernel).
+/// Zero-initialize the structure, including currently unused fields, to
+/// accommodate potential future updates.
+
+#[derive(Debug)]
+#[repr(C)]
+pub(crate) struct SpiIocTransfer {
+    tx_buf: u64,
+    rx_buf: u64,
+    len: u32,
+    speed_hz: u32,
+    delay_usecs: u16,
+    bits_per_word: u8,
+    cs_change: u8,
+    tx_nbits: u8,
+    rx_nbits: u8,
+    word_delay_usecs: u8,
+    pad: u8,
+}
+
+/// SPI definitions
+pub(crate) struct SpiTransReq {
+    pub tx_buf: Vec<u8>,
+    pub rx_buf: Vec<u8>,
+    pub trans_len: u32,
+    pub speed_hz: u32,
+    pub mode: u32,
+    pub delay_usecs: u16,
+    pub bits_per_word: u8,
+    pub cs_change: u8,
+    pub tx_nbits: u8,
+    pub rx_nbits: u8,
+    pub word_delay_usecs: u8,
+    pub cs_id: u8,
+}
+
+/// Virtio SPI Configuration
+#[derive(Copy, Clone, Debug, Default, PartialEq)]
+#[repr(C)]
+pub(crate) struct VirtioSpiConfig {
+    pub(crate) cs_max_number: u8,
+    pub(crate) cs_change_supported: u8,
+    pub(crate) tx_nbits_supported: u8,
+    pub(crate) rx_nbits_supported: u8,
+    pub(crate) bits_per_word_mask: Le32,
+    pub(crate) mode_func_supported: Le32,
+    pub(crate) max_freq_hz: Le32,
+    pub(crate) max_word_delay_ns: Le32,
+    pub(crate) max_cs_setup_ns: Le32,
+    pub(crate) max_cs_hold_ns: Le32,
+    pub(crate) max_cs_inactive_ns: Le32,
+}
```

Comment:
Here I reference the vhost-device-gpio, in which the virtio config is also defined in gpio.rs, since it's a member of GpioController. 
In Spi, I set VirtioSpiConfig as the SpiController's member, and the initialization of VirtioSpiConfig also depends on the Linux Spi access. Do you think this is acceptable here? Looking forward to your advice. Thanks a lot.

Reply:
Though I have written GPIO backend too, I feel that structure is wrongly placed there. Can be moved to vhu_gpio.rs..